### PR TITLE
refactor!: replace `ValueReceived` with `UniversalReceiver`

### PR DESCRIPTION
**Pull request by YamenMerhi**
# What does this PR introduce?

## ♻️ Refactor

- Emitting UniversalReceiver event instead of ValueReceived in owner functions (setData, execute, ..)
- Emitting UniversalReceiver event instead of ValueReceived and reacting on the call using UniversalReceiverDelegates in the `receive`, `fallback` and the `universalReceiver` function.
- Update `LSP17` to optionally forward value in `LSP0` and `LSP9`

### PR Checklist

- [ ] Wrote Tests
- [ ] Wrote & Generated Documentation (readme/natspec/dodoc)
- [ ] Ran `npm run lint` && `npm run lint:solidity` (solhint)
- [ ] Ran `npm run format` (prettier)
- [ ] Ran `npm run build`
- [ ] Ran `npm run test`

### CODE REVIEW

**Review by b00ste — APPROVED** [review]

File: tests/UniversalProfile.behaviour.ts [hunk 1]
```diff
@@ -559,6 +594,148 @@ export const shouldBehaveLikeLSP3 = (
       });
     });
   });
+
+  describe('when setting a UniversalReceiverDelegate for typeId of LYX receiving', () => {
```
Comment (CJ42) [inline tests/UniversalProfile.behaviour.ts:598]: suggestion

```suggestion
  describe('when setting a UniversalReceiverDelegate for typeId `LSP0ValueReceived`', () => {
```

File: tests/UniversalProfile.behaviour.ts [hunk 2]
```diff
@@ -559,6 +594,148 @@ export const shouldBehaveLikeLSP3 = (
...
+      });
+    });
+
+    describe('when sending empty call to the receive function', () => {
+      it('should not react on the call and not emit UniversalReceiver', async () => {
+        const tx = await context.accounts[1].sendTransaction({
+          to: context.universalProfile.address,
+        });
+
+        expect(tx).to.not.emit(context.universalProfile, 'UniversalReceiver');
```
Comment (CJ42) [inline tests/UniversalProfile.behaviour.ts:637]: amazing, great not test 👌

**Review by CJ42 — APPROVED** [review]

File: contracts/LSP17ContractExtension/LSP17Extendable.sol [hunk 3]
```diff
@@ -90,15 +91,18 @@ abstract contract LSP17Extendable is ERC165 {
         bytes calldata callData
     ) internal virtual returns (bytes memory) {
         // If there is a function selector
-        address extension = _getExtension(msg.sig);
+        (
+            address extension,
+            bool isForwardingValue
```
[the remark this answers is not in the document]
Comment (CJ42) [reply contracts/LSP17ContractExtension/LSP17Extendable.sol:96]: We keep it as it is as decided

[the remark this answers is not in the document]
Comment (CJ42) [reply contracts/LSP17ContractExtension/LSP17Extendable.sol:96]: Finally we apply the suggested changes.

**Review by CJ42 — APPROVED** [review]

File: contracts/LSP0ERC725Account/LSP0ERC725AccountCore.sol [hunk 4]
```diff
@@ -791,21 +823,29 @@ abstract contract LSP0ERC725AccountCore is
...
-        // Check if there is an extension stored under the generated data key
-        address extension = address(
-            bytes20(ERC725YCore._getData(mappedExtensionDataKey))
+        bytes memory extensionData = ERC725YCore._getData(
+            mappedExtensionDataKey
         );
 
-        return extension;
+        // Check if the extensionData is 21 bytes long (20 bytes of address + 1 byte as bool indicator to forwards the value)
+        if (extensionData.length == 21) {
```
[the remark this answers is not in the document]
Comment (CJ42) [reply contracts/LSP0ERC725Account/LSP0ERC725AccountCore.sol:842]: Decided to keep as it was

**Comment by CJ42** [comment]
We wil implement the suggested changes for the tests in a separate PR

**Pull request merged**